go/runtime/host/loadbalance/loadbalance.go

```diff
@@ -0,0 +1,218 @@
+)
+
+// Config is the configuration for the runtime load balancer.
+type Config struct {
+	// NumInstances is the number of runtime instances to provision.
+	NumInstances int
+}
+
+type lbRuntime struct {
+	id        common.Namespace
+	instances []host.Runtime
+
+	lastIdx atomic.Uint64
+}
+
+// Implements host.Runtime.
+func (lb *lbRuntime) ID() common.Namespace {
+	return lb.id
+}
+
+// Implements host.Runtime.
+func (lb *lbRuntime) GetInfo(ctx context.Context) (*protocol.RuntimeInfoResponse, error) {
+	return lb.instances[0].GetInfo(ctx)
+}
+
+// Implements host.Runtime.
+func (lb *lbRuntime) GetCapabilityTEE() (*node.CapabilityTEE, error) {
+	// TODO: This won't work when registration of all client runtimes is required.
+	return lb.instances[0].GetCapabilityTEE()
+}
+
+// shouldPropagateToAll checks whether the given runtime request should be propagated to all
+// instances.
+func shouldPropagateToAll(body *protocol.Body) bool {
+	switch {
+	case body.RuntimeConsensusSyncRequest != nil:
+		// Consensus view of all instances should be up to date as otherwise signed attestations
+		// will be stale, resulting in them being rejected.
+		return true
+	case body.RuntimeKeyManagerPolicyUpdateRequest != nil,
+		body.RuntimeKeyManagerStatusUpdateRequest != nil,
+		body.RuntimeKeyManagerQuotePolicyUpdateRequest != nil:
+		// Key manager updates should be propagated.
+		return true
+	default:
+		return false
+	}
+}
+
+// Implements host.Runtime.
+func (lb *lbRuntime) Call(ctx context.Context, body *protocol.Body) (*protocol.Body, error) {
+	switch {
+	case shouldPropagateToAll(body):
+		// Propagate call to all instances.
+		type result struct {
+			rsp *protocol.Body
+			err error
+		}
+		resCh := make(chan *result, len(lb.instances))
+		for _, rt := range lb.instances {
+			rt := rt // Make sure goroutine below operates on the right instance.
+
+			go func() {
+				rsp, err := rt.Call(ctx, body)
+				resCh <- &result{
+					rsp: rsp,
+					err: err,
+				}
+			}()
+		}
+
+		var (
+			anyErr error
+			rsp    *protocol.Body
+		)
+		for range lb.instances {
+			res := <-resCh
+			if res.err == nil {
+				rsp = res.rsp
+			} else {
+				anyErr = errors.Join(anyErr, res.err)
+			}
+		}
+		return rsp, anyErr
+	case body.RuntimeQueryRequest != nil, body.RuntimeCheckTxBatchRequest != nil:
+		// Load-balance queries.
+		idx := lb.lastIdx.Add(1) % uint64(len(lb.instances))
+
+		lbRequestCount.With(prometheus.Labels{
+			"runtime":     lb.id.String(),
+			"lb_instance": fmt.Sprintf("%d", idx),
+		}).Inc()
+
+		return lb.instances[int(idx)].Call(ctx, body)
+	default:
+		// Propagate only to the first instance.
+		return lb.instances[0].Call(ctx, body)
+	}
+}
+
+// Implements host.Runtime.
+func (lb *lbRuntime) UpdateCapabilityTEE() {
+	for _, rt := range lb.instances {
+		rt.UpdateCapabilityTEE()
+	}
+}
+
+// Implements host.Runtime.
+func (lb *lbRuntime) WatchEvents(ctx context.Context) (<-chan *host.Event, pubsub.ClosableSubscription, error) {
```

Comment:
Is it ok to not handle FailedToStart/Stopped events of other runtimes? 
Can it happen that for some reason some (non-first) runtimes silently stop and won't be able to serve requests anymore while the LB will still be sending them requests? (should the LB handle that and try restarting those runtimes?)

Reply:
The regular provisioner already automatically handles runtime restarts if they fail.

But marking instances as failed and not routing requests to them indeed makes sense.